psh/cd/cd.c

```diff
@@ -0,0 +1,114 @@
+/*
+ * Phoenix-RTOS
+ *
+ * cd - changes working directory
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "../psh.h"
+
+
+static void psh_cd_info(void)
+{
+	printf("changes the working directory");
+}
+
+
+static void psh_cd_help(const char *progname)
+{
+	printf("Usage: %s [directory]\n", progname);
+}
+
+
+static const char *psh_strerror(int errn)
+{
+	switch (errno) {
+		case ENOENT: return "No such file or directory";
+		case ENOTDIR: return "Not a directory";
+		default: break;
+	}
+
+	return strerror(errno);
+}
+
+
+static int do_chdir(const char *dst)
+{
+	int res, print = 0;
+	char *cwd = getcwd(NULL, 0);
+
+	/* TODO: implement the search of the CDPATH if needed */
+
+	if (dst == NULL) {
+		dst = getenv("HOME");
+		if (dst == NULL || dst[0] == '\0') {
+			dst = "/";
+		}
+	}
+	else if (dst[0] == '-' && dst[1] == '\0') {
+		print = 1;
+		dst = getenv("OLDPWD");
+		if (dst == NULL) {
+			fprintf(stderr, "cd: OLDPWD has not yet been set\n");
+			return -1;
+		}
+	}
+
+	res = chdir(dst);
+	if (res < 0) {
+		fprintf(stderr, "cd: %s: %s\n", dst, psh_strerror(errno));
+		return -1;
+	}
+
+	if (print) {
+		puts(dst);
+	}
+
+	if (cwd != NULL) {
+		setenv("OLDPWD", cwd, 1);
+		free(cwd);
+	}
+
+	return EOK;
+}
+
+
+static int psh_cd(int argc, char **argv)
+{
+	int res = -1;
+
+	if (argc == 2) {
+		if (strcmp(argv[1], "-h") == 0) {
+			psh_cd_help(argv[0]);
+			return EXIT_SUCCESS;
+		}
+		res = do_chdir(argv[1]);
+	}
+	else if (argc == 1) {
+		res = do_chdir(NULL);
+	}
+	else {
+		psh_cd_help(argv[0]);
+	}
```

Comment:
It's good to use getopt anyway, there's a good chance we'll update applet functionality in the future, also it handles wierd use-cases better
e.g. cd -h -h, cd dir -opt

Reply:
Ok, I've replaced it with `getopt()`, I put the `switch`-`case` instead of single `if` as a placeholder since I will add support for `-P` and `-L` switches here soon.